internal/authmiddleware/oidc_verifier.go

```diff
@@ -0,0 +1,217 @@
+		"issuer", config.OIDCIssuerURL,
+		"client_id", config.OIDCClientID,
+		"timeout_secs", config.OIDCInitTimeoutSecs,
+	)
+
+	// Create the token verifier
+	// Note that any caller may obtain the public keys from the OIDC well-known page, a client ID
+	// and client secret are only needed to issue new token (and secret are optional for public
+	// access points).
+	// In authmiddleware flow, the OIDC provider will issue token to a different component (e.g. oauth2-proxy)
+	// and such component will pass the token as Authentication token to authmiddleware,
+	// thus we use the client ID of the previous component as client.
+	oidcConfig := &oidc.Config{
+		ClientID:          config.OIDCClientID,
+		SkipClientIDCheck: false,
+	}
+
+	return &OIDCVerifier{
+		provider:       nil, // Will be initialized in Start()
+		verifier:       nil, // Will be initialized in Start()
+		clientID:       config.OIDCClientID,
+		issuerURL:      config.OIDCIssuerURL,
+		logger:         logger,
+		timeoutSeconds: config.OIDCInitTimeoutSecs,
+		oidcConfig:     oidcConfig,
+	}, nil
+}
+
+// Start initializes the OIDC provider and verifier
+// This allows deferring HTTP calls until the application is ready
+func (v *OIDCVerifier) Start(ctx context.Context) error {
+	if v.provider != nil {
+		// Already initialized
+		return nil
+	}
+
+	v.logger.Info("Starting OIDC verifier - initializing provider connection",
+		"issuer", v.issuerURL,
+		"client_id", v.clientID,
+	)
+
+	// Create a context with timeout for OIDC provider initialization
+	initCtx, cancel := context.WithTimeout(ctx, time.Duration(v.timeoutSeconds)*time.Second)
+	defer cancel()
+
+	// Initialize OIDC provider and verifier - this makes HTTP calls
+	if v.logger != nil {
+		v.logger.Info("Configuring new OIDC provider", "issuerURL", v.issuerURL)
+	}
+	provider, err := oidc.NewProvider(initCtx, v.issuerURL)
+	if err != nil {
+		return fmt.Errorf("failed to initialize OIDC provider: %w", err)
+	}
+	v.provider = provider
+	if v.logger != nil {
+		v.logger.Info("OIDC provider is ready")
+	}
+
+	if v.logger != nil {
+		v.logger.Info("Configuring token verifier",
+			"issuer URL", v.issuerURL,
+			"client ID", v.clientID)
+	}
+	v.verifier = provider.Verifier(v.oidcConfig)
+	if v.logger != nil {
+		v.logger.Info("Token verifier is ready")
+	}
+	return nil
+}
+
+// VerifyToken verifies an OIDC token and returns Claims, isFault, error.
+// It may call the provider to refresh the public keySet if not cached
+func (v *OIDCVerifier) VerifyToken(ctx context.Context, tokenString string, logger *slog.Logger) (*OIDCClaims, bool, error) {
+	if v.verifier == nil {
+		return nil, true, fmt.Errorf("OIDC verifier is not initialized - call Start() first")
+	}
+
+	// Verify the token
+	idToken, err := v.verifier.Verify(ctx, tokenString)
+	if err != nil {
+		// Check if this is a discovery document error
+		errMsg := err.Error()
+		if strings.Contains(errMsg, "failed to get discovery document") ||
```

Comment:
Should error type assertions be used here instead of string matching to classify errors as string matching is / can be fragile?